File: app/java/app/src/test/java/com/example/accounting/AccountRefactoringTest.java
package com.example.accounting;

import com.example.accounting.testsetup.TestDatabase;
import org.junit.jupiter.api.*;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.math.BigDecimal;
import java.sql.*;

import static org.assertj.core.api.Assertions.*;

/**
 * 勘定科目マスタのリファクタリングテスト
 *
 * 実務で必要な項目（BSPL区分、取引要素区分、合計科目フラグ等）を段階的に追加します。
 */
@Testcontainers
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class AccountRefactoringTest {

    @Container
    static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>("postgres:16-alpine")
            .withDatabaseName("testdb")
            .withUsername("testuser")
            .withPassword("testpass");

    private static TestDatabase testDb;
    private static Connection connection;

    @BeforeAll
    static void setUp() throws SQLException {
        testDb = new TestDatabase(postgres);
        testDb.start();
        connection = testDb.getConnection();
    }

    @AfterAll
    static void tearDown() throws SQLException {
        testDb.stop();
    }

    @BeforeEach
    void cleanup() throws SQLException {
        testDb.cleanup();
    }

    @Test
    @Order(1)
    @DisplayName("BSPL区分を設定できる")
    void testBsplDistinction() throws SQLException {
        // マイグレーション後、このテストが通るようになる
        String sql = """
            INSERT INTO "勘定科目マスタ"
            ("勘定科目コード", "勘定科目名", "勘定科目種別", "BSPL区分", "残高")
            VALUES (?, ?, ?::account_type, ?, ?)
            RETURNING "勘定科目コード", "BSPL区分"
            """;

        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            pstmt.setString(1, "1000");
            pstmt.setString(2, "現金");
            pstmt.setString(3, "資産");
            pstmt.setString(4, "B");  // 貸借対照表
            pstmt.setBigDecimal(5, new BigDecimal("0"));

            ResultSet rs = pstmt.executeQuery();

            assertThat(rs.next()).isTrue();
            assertThat(rs.getString("BSPL区分")).isEqualTo("B");
        }
    }

    @Test
    @Order(2)
    @DisplayName("取引要素区分を設定できる")
    void testTransactionDistinction() throws SQLException {
        String sql = """
            INSERT INTO "勘定科目マスタ"
            ("勘定科目コード", "勘定科目名", "勘定科目種別", "BSPL区分", "取引要素区分", "残高")
            VALUES (?, ?, ?::account_type, ?, ?, ?)
            RETURNING "勘定科目コード", "取引要素区分"
            """;

        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            pstmt.setString(1, "1000");
            pstmt.setString(2, "現金");
            pstmt.setString(3, "資産");
            pstmt.setString(4, "B");
            pstmt.setString(5, "1");  // 資産
            pstmt.setBigDecimal(6, new BigDecimal("0"));

            ResultSet rs = pstmt.executeQuery();

            assertThat(rs.next()).isTrue();
            assertThat(rs.getString("取引要素区分")).isEqualTo("1");
        }
    }

    @Test
    @Order(3)
    @DisplayName("集計科目を設定できる")
    void testSumAccount() throws SQLException {
        String sql = """
            INSERT INTO "勘定科目マスタ"
            ("勘定科目コード", "勘定科目名", "勘定科目種別", "合計科目", "残高")
            VALUES (?, ?, ?::account_type, ?, ?)
            RETURNING "勘定科目コード", "合計科目"
            """;

        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            pstmt.setString(1, "1000");
            pstmt.setString(2, "流動資産");
            pstmt.setString(3, "資産");
            pstmt.setBoolean(4, true);  // 集計科目フラグ
            pstmt.setBigDecimal(5, new BigDecimal("0"));

            ResultSet rs = pstmt.executeQuery();

            assertThat(rs.next()).isTrue();
            assertThat(rs.getBoolean("合計科目")).isTrue();
        }
    }

    @Test
    @Order(4)
    @DisplayName("勘定科目カナを設定できる")
    void testAccountKana() throws SQLException {
        String sql = """
            INSERT INTO "勘定科目マスタ"
            ("勘定科目コード", "勘定科目名", "勘定科目種別", "勘定科目カナ", "残高")
            VALUES (?, ?, ?::account_type, ?, ?)
            RETURNING "勘定科目コード", "勘定科目カナ"
            """;

        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            pstmt.setString(1, "1000");
            pstmt.setString(2, "現金");
            pstmt.setString(3, "資産");
            pstmt.setString(4, "ゲンキン");
            pstmt.setBigDecimal(5, new BigDecimal("0"));

            ResultSet rs = pstmt.executeQuery();

            assertThat(rs.next()).isTrue();
            assertThat(rs.getString("勘定科目カナ")).isEqualTo("ゲンキン");
        }
    }

    @Test
    @Order(5)
    @DisplayName("表示順序と集計対象を設定できる")
    void testDisplayOrderAndAggregation() throws SQLException {
        String sql = """
            INSERT INTO "勘定科目マスタ"
            ("勘定科目コード", "勘定科目名", "勘定科目種別", "表示順序", "集計対象", "残高")
            VALUES (?, ?, ?::account_type, ?, ?, ?)
            RETURNING "勘定科目コード", "表示順序", "集計対象"
            """;

        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            pstmt.setString(1, "1000");
            pstmt.setString(2, "現金");
            pstmt.setString(3, "資産");
            pstmt.setInt(4, 100);
            pstmt.setBoolean(5, true);
            pstmt.setBigDecimal(6, new BigDecimal("0"));

            ResultSet rs = pstmt.executeQuery();

            assertThat(rs.next()).isTrue();
            assertThat(rs.getInt("表示順序")).isEqualTo(100);
            assertThat(rs.getBoolean("集計対象")).isTrue();
        }
    }
}

File: app/csharp/AccountingSystem.Tests/AccountRefactoringTest.cs
using Xunit;
using FluentAssertions;
using Npgsql;
using Testcontainers.PostgreSql;

namespace AccountingSystem.Tests
{
    /// <summary>
    /// 勘定科目マスタのリファクタリングテスト
    /// </summary>
    public class AccountRefactoringTest : IAsyncLifetime
    {
        private readonly PostgreSqlContainer _postgres;
        private TestDatabase? _testDb;
        private NpgsqlConnection? _connection;

        public AccountRefactoringTest()
        {
            _postgres = new PostgreSqlBuilder()
                .WithImage("postgres:16-alpine")
                .WithDatabase("testdb")
                .WithUsername("testuser")
                .WithPassword("testpass")
                .Build();
        }

        public async Task InitializeAsync()
        {
            await _postgres.StartAsync();
            _testDb = new TestDatabase(_postgres);
            await _testDb.StartAsync();
            _connection = _testDb.GetConnection();
        }

        public async Task DisposeAsync()
        {
            if (_testDb != null)
            {
                await _testDb.StopAsync();
            }
            await _postgres.DisposeAsync();
        }

        private async Task CleanupAsync()
        {
            await _testDb!.CleanupAsync();
        }

        [Fact(DisplayName = "BSPL区分を設定できる")]
        public async Task TestBsplDistinction()
        {
            await CleanupAsync();

            // マイグレーション後、このテストが通るようになる
            var sql = @"
                INSERT INTO ""勘定科目マスタ""
                (""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""BSPL区分"", ""残高"")
                VALUES (@code, @name, @type::account_type, @bspl, @balance)
                RETURNING ""勘定科目コード"", ""BSPL区分""
            ";

            await using var cmd = new NpgsqlCommand(sql, _connection);
            cmd.Parameters.AddWithValue("code", "1000");
            cmd.Parameters.AddWithValue("name", "現金");
            cmd.Parameters.AddWithValue("type", "資産");
            cmd.Parameters.AddWithValue("bspl", "B");  // 貸借対照表
            cmd.Parameters.AddWithValue("balance", 0m);

            await using var reader = await cmd.ExecuteReaderAsync();

            (await reader.ReadAsync()).Should().BeTrue();
            reader.GetString(reader.GetOrdinal("BSPL区分")).Should().Be("B");
        }

        [Fact(DisplayName = "取引要素区分を設定できる")]
        public async Task TestTransactionDistinction()
        {
            await CleanupAsync();

            var sql = @"
                INSERT INTO ""勘定科目マスタ""
                (""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""BSPL区分"", ""取引要素区分"", ""残高"")
                VALUES (@code, @name, @type::account_type, @bspl, @trans, @balance)
                RETURNING ""勘定科目コード"", ""取引要素区分""
            ";

            await using var cmd = new NpgsqlCommand(sql, _connection);
            cmd.Parameters.AddWithValue("code", "1000");
            cmd.Parameters.AddWithValue("name", "現金");
            cmd.Parameters.AddWithValue("type", "資産");
            cmd.Parameters.AddWithValue("bspl", "B");
            cmd.Parameters.AddWithValue("trans", "1");  // 資産
            cmd.Parameters.AddWithValue("balance", 0m);

            await using var reader = await cmd.ExecuteReaderAsync();

            (await reader.ReadAsync()).Should().BeTrue();
            reader.GetString(reader.GetOrdinal("取引要素区分")).Should().Be("1");
        }

        [Fact(DisplayName = "合計科目を設定できる")]
        public async Task TestSumAccount()
        {
            await CleanupAsync();

            var sql = @"
                INSERT INTO ""勘定科目マスタ""
                (""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""合計科目"", ""残高"")
                VALUES (@code, @name, @type::account_type, @sum, @balance)
                RETURNING ""勘定科目コード"", ""合計科目""
            ";

            await using var cmd = new NpgsqlCommand(sql, _connection);
            cmd.Parameters.AddWithValue("code", "1000");
            cmd.Parameters.AddWithValue("name", "流動資産");
            cmd.Parameters.AddWithValue("type", "資産");
            cmd.Parameters.AddWithValue("sum", true);  // 集計科目フラグ
            cmd.Parameters.AddWithValue("balance", 0m);

            await using var reader = await cmd.ExecuteReaderAsync();

            (await reader.ReadAsync()).Should().BeTrue();
            reader.GetBoolean(reader.GetOrdinal("合計科目")).Should().BeTrue();
        }
    }
}
